src/Microsoft.TemplateSearch.TemplateDiscovery/Nuget/NugetPackProvider.cs

```diff
@@ -25,11 +30,21 @@ internal NugetPackProvider(string name, string query, string packageTempBasePath
             _pageSize = pageSize;
             _runOnlyOnePage = runOnlyOnePage;
             _packageTempPath = Path.GetFullPath(Path.Combine(packageTempBasePath, DownloadedPacksDir, Name));
-            _searchUriFormat = $"https://api-v2v3search-0.nuget.org/query?{query}&skip={{0}}&take={{1}}&prerelease={includePreviewPacks}";
+            _repository = Repository.Factory.GetCoreV3(NuGetOrgFeed);
+            ServiceIndexResourceV3 indexResource = _repository.GetResource<ServiceIndexResourceV3>();
+            IReadOnlyList<ServiceIndexEntry> searchResources = indexResource.GetServiceEntries("SearchQueryService");
+            _downloadResource = _repository.GetResource<FindPackageByIdResource>();
+
+            if (!searchResources.Any())
+            {
+                throw new Exception($"{NuGetOrgFeed} does not support search API (SearchQueryService)");
+            }
+
+            _searchUriFormat = $"{searchResources[0].Uri}?{query}&skip={{0}}&take={{1}}&prerelease={includePreviewPacks}";
 
             if (Directory.Exists(_packageTempPath))
             {
-                throw new Exception($"temp storage path for nuget packages already exists");
+                throw new Exception($"temp storage path for NuGet packages already exists: {_packageTempPath}");
```

Comment:
minor: consider using an Exception type that closely resembles the encountered problem.  Perhaps InvalidOperation.